KEP-3331: add audienceMatchPolicy to authentication config (#4430)
kubernetes/enhancements: keps/sig-auth/3331-structured-authentication-configuration/README.md

        CertificateAuthority string `json:"certificateAuthority,omitempty"`

        // audiences is the set of acceptable audiences the JWT must be issued to.
        // Same value as the --oidc-client-id flag (though this field supports an array).
        // Required to be non-empty.
        Audiences []string `json:"audiences,omitempty"`

        // audienceMatchPolicy defines how the "Audiences" field is used to match the "aud" claim in the presented JWT.
        // Allowed values are "MatchAny" or "MatchAll" when multiple audiences are specified and
        // empty when a single audience is specified.
        //

>         // - MatchAny: the "aud" claim in the presented JWT must match at least one of the entries in the "Audiences" field.

Reviewer: ```suggestion
        // - MatchAny: the "aud" claim in the presented JWT must match at least one of the entries in the "audiences" field.
```

>         // - MatchAny: the "aud" claim in the presented JWT must match at least one of the entries in the "audiences" field.

Reviewer 2: MatchAny is the way I'd expect a server to be configured if it has multiple ways it can be addressed, and wants to accept a JWT for any of those URLs. There are clear reasons for this option.

>         // For example, if "Audiences" is ["foo", "bar"], the "aud" claim in the presented JWT must contain either "foo" or "bar".

Reviewer: ```suggestion
        // For example, if "audiences" is ["foo", "bar"], the "aud" claim in the presented JWT must contain either "foo" or "bar" (and may contain both).
```

        //

>         // - MatchAll: the "aud" claim in the presented JWT must match all of the entries in the "audiences" field.

Reviewer 2: hoisting my comment from slack:

> match_all is a stretch from normal jwt semantics
>
> normally jwts can have multiple audiences and a particular server matches one (edited) 
>
> I can see a case for match_any where a server has an "internal" audience and "external" audience
>
> match_all ("this JWT MUST have multiple audiences and must have ALL of these audiences") is really weird

What is the scenario where MatchAll would make sense, and does that usage fit the meaning of the audience claim?

Author: If the plan is to only support `MatchAny`, is there a necessity for the introduction of the new field? Perhaps we could revise the documentation for the `Audiences` field to explicitly detail this and recommend users to use claim validation rules for intricate matching scenarios.

Reviewer: With the `audienceMatchPolicy` field we get a direct opt-in by the user.  Unclear if that is enough justification for the field.

Reviewer 2: I still don't see a good reason for MatchAll, but to be explicit when we have more than one audience in the config seems fine:

* If `audiences` has a single item, allow audienceMatchPolicy to be "" or "MatchAny", and either option keeps the current behavior (audience must be present in the JWT, additional audiences in the JWT are tolerated/ignored).

* If `audiences` has multiple items, require audienceMatchPolicy to be set, the only currently allowed value is "MatchAny", and we require at least one of the configured audiences be present in the JWT, additional audiences in the JWT are tolerated/ignored).

If we get compelling use cases in the future for other things like MatchAll or MatchNone or MatchAnyStrict (to get behavior like https://github.com/kubernetes/enhancements/pull/4430/files#r1469912712), this gives us a way to wire those in as well

Author: updated!

        // For example:
        // 1. If "Audiences" is ["foo", "bar"], the "aud" claim in the presented JWT must contain both "foo" and "bar".
        // 2. If "Audiences" is ["foo", "bar"], the "aud" claim in the presented JWT is ["foo", "bar", "baz"], the token is valid.
        //

>         // - "": The match policy should be empty when a single audience is specified in the "Audiences" field.

Reviewer: ```suggestion
        // - "": The match policy must be empty (or unset) when a single audience is specified in the "audiences" field.
```
